### PR TITLE
Refactor/virtual machine

### PR DESCRIPTION
**Pull request by dpl0a**
This pull request packages the data passed to the evaluation functions (e.g. `eval`, `eval_closure`, ecc) in a `VirtualMachine` struct and moves the evaluation functions as part of the implementation for `VirtualMachine`.

### CODE REVIEW

File: src/repl/mod.rs [hunk 1]
```diff
@@ -177,7 +177,7 @@ impl ReplImpl {
         match term {
             ExtendedTerm::RichTerm(t) => {
                 let t = prepare(self, None, t)?;
-                Ok(eval_function(t, &self.env.eval_env, &mut self.cache)?.into())
+                Ok(eval_function(&mut eval::VirtualMachine::new(&mut self.cache), t, &self.env.eval_env)?.into())
```
Comment (yannham) [inline src/repl/mod.rs:180]: That's not very important, but in the REPL, we are evaluating things repeatedly. I wonder if that wouldn't be a tad cleaner to store a `vm` in the state of the REPL and to reuse it for each evaluation, with a prior call to `reset()`

**Review by yannham — APPROVED** [review]
God, git diff are really bad at handling code moves. This is not really reviewable everywhere, but the changes in `eval::mod` and `eval::operation` were mostly mechanical, so I trust they are correct. I left a bunch of non-blocking comments: you can decide what to do with them (including ignoring them), and then merge once you are satisfied (and the CI is green).

File: src/eval/mod.rs [hunk 2]
```diff
@@ -119,612 +119,627 @@ impl AsRef<Vec<StackElem>> for CallStack {
...
 
-/// Kind of an identifier.
-#[derive(Debug, PartialEq, Eq, Clone, Copy)]
-pub enum IdentKind {
-    Let,
-    Lambda,
-    Record,
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum EvalMode {
+    UnwrapMetaValues,
```
[the remark this answers is not in the document]
Comment (dpl0a) [reply src/eval/mod.rs:124]: UnwrapMeta and StopAtMeta maybe?

File: src/repl/mod.rs (hunk 1 above)
Comment (yannham) [reply src/repl/mod.rs:180]: Sounds good (small PRs are always good :slightly_smiling_face: )

File: src/eval/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (yannham) [reply src/eval/mod.rs:124]: The StopAtMeta doesn't sound glorious, but well, at least it's descriptive, so let's go with it for now.

File: src/eval/mod.rs [hunk 3]
```diff
@@ -119,612 +119,627 @@ impl AsRef<Vec<StackElem>> for CallStack {
...
     }
 }
 
-pub type Environment = GenericEnvironment<Ident, Thunk>;
-
-/// Raised when trying to build an environment from a term which is not a record.
-#[derive(Clone, Debug)]
-pub enum EnvBuildError {
-    NotARecord(RichTerm),
+pub struct VirtualMachine<'a, R: ImportResolver> {
```
[the remark this answers is not in the document]
Comment (dpl0a) [reply src/eval/mod.rs:134]: The thing is, `VirtualMachine` was already kind of documented before existing

```rust
//! # The Nickel Abstract Machine
//! The abstract machine is a stack machine composed of the following elements:
//! - The term being currently evaluated
//! - The main stack, storing arguments, thunks and pending computations
//! - A pair of [environment][Environment], mapping identifiers to [closures][Closure]:
//!     * The initial environment contains builtin functions accessible from anywhere, and alive
//!     during the whole evaluation
//!     * The local environment contains the variables in scope of the current term and is subject
//!     to garbage collection (currently reference counting based)
//! - A [callstack], mainly for error reporting purpose
```
This is lines 7-16 in eval/mod.rs
Should I move/copy the relevant part of this description?

**Pull request merged**